### PR TITLE
Use custom favicon when viewing static files if it exists

### LINKED ISSUE #19109: Favicon not used when viewing files
### Gitea Version

1.15.9

### Operating System

FreeBSD 13

### Browser Version

Firefox 98.0

### Can you reproduce the bug on the Gitea demo site?

No

### Description

Cannot reproduce at try.gitea.io because it does not use a non-standard favicon. See [this file on my self-hosted Gitea instance as an example](https://git.fwdekker.com/avatars/63edb731af35212d057736d0ff52fe6b).

When looking at a file directly in the browser, such as an avatar or PDF file, the default favicon is used instead of the custom favicon. At least in Firefox, the browser displays the favicon at `example.com/favicon.ico` when viewing a file. On HTML pages, the browser correctly uses `example.com/assets/img/logo.svg`.

I have configured the favicon by adding the files `favicon.ico`, `favicon.png`, `logo.png`, and `logo.svg` to `$GITEA_CUSTOM/public/img/`.

### Screenshots

Two tabs showing different favicons for the same Gitea instance.

On the left: correct favicon is used for an HTML page.
On the right: default favicon is used for an image.

### PR DESCRIPTION
**Pull request by abheekda1**
Redirect `/favicon.ico` to `/assets/img/favicon.png`.

Fix #19109

### CODE REVIEW

File: routers/web/web.go [hunk 1]
```diff
@@ -99,6 +99,18 @@ func Routes(sessioner func(http.Handler) http.Handler) *web.Route {
 		http.Redirect(w, req, path.Join(setting.StaticURLPrefix, "/assets/img/apple-touch-icon.png"), 301)
 	})
 
+	// the custom favicon should be seen when viewing files instead of the default one if it exists,
+	routes.Get("/favicon.ico", func(w http.ResponseWriter, req *http.Request) {
+		faviconPath := ""
+		res, err := http.Get(path.Join(setting.StaticURLPrefix, "/assets/img/favicon.ico"))
```
Comment (Gusted) [inline routers/web/web.go]: It's better to only check this once and not make a request every time when `favicon.ico` is requested. We can use `os.Stat` to check if such file exists.

**Comment by ADawesomeguy** [comment]
Alright, cool! I'll give it a shot.

File: routers/web/web.go (hunk 1 above)
Comment (silverwind) [reply routers/web/web.go]: There are a number of logo files that are known to exist, basically the output files here:

https://github.com/go-gitea/gitea/blob/fb08d2b3fd04dc8fc6cbd45fd341773b817c0857/build/generate-images.js#L74-L78

If browser accept redirecting ico to either svg or png, I think this is the solution we should go for as `favicon.ico` generally does not exist for our installations (but users could provide it for compatibilty)

Comment (abheekda1) [reply routers/web/web.go]: I think it may work to redirect to PNG, I'll give that a shot.

Comment (silverwind) [reply routers/web/web.go]: Yeah, it's likely if PNG works, it will work with SVG as well, see https://caniuse.com/link-icon-svg for browser support.

Comment (silverwind) [reply routers/web/web.go]: Another option might be to just generate a `ico` in above script, but that might be more complicated, and I think it should only be a last-resort option if those redirects don't work.

**Review by zeripath — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by Gusted — APPROVED** [review]

**Comment by zeripath** [comment]
make lgtm work

**Comment by zeripath** [comment]
I guess we should backport this.

**Comment by zeripath** [comment]
@ADawesomeguy are you able to send a backport PR?


<details>
<summary>This is my backport script: (if it helps)</summary>

```sh
#!/bin/sh
PR="$1"
SHA="$2"
VERSION="$3"

if [ -z "$SHA" ]; then
    SHA=$(gh api /repos/go-gitea/gitea/pulls/$PR -q '.merge_commit_sha')
fi

if [ -z "$VERSION" ]; then
    VERSION="v1.16"
fi

echo git checkout origin/release/"$VERSION" -b backport-$PR-$VERSION
git checkout origin/release/"$VERSION" -b backport-$PR-$VERSION
git cherry-pick $SHA && git commit --amend && git push ADawesomeguy backport-$PR-$VERSION && xdg-open https://github.com/go-gitea/gitea/compare/release/"$VERSION"...ADawesomeguy:backport-$PR-$VERSION

```

</details>

**Comment by ADawesomeguy** [comment]
Do you want me to run that, because yeah sure I can do that. What exactly is a backport?

**Comment by zeripath** [comment]
By backport I mean cherry-pick f96e8be41fe2cc36858bd4a673847bb388ce178a on to release/v1.16,

That cherrypick branch would then be used to open a PR against release/v1.16 titled: `Use custom favicon when viewing static files if it exists (#19130)` (← it must be this title) the PR initial comment should be:

```
Backport #19130

Redirect `/favicon.ico` to `/assets/img/favicon.png`.

Fix #19109 
```

**Comment by ADawesomeguy** [comment]
@zeripath like this? https://github.com/ADawesomeguy/gitea/commits/backport-19130-v1.16

**Comment by zeripath** [comment]
https://github.com/go-gitea/gitea/compare/release/v1.16...ADawesomeguy:backport-19130-v1.16

**Comment by ADawesomeguy** [comment]
Alright, done!

**Comment by silverwind** [comment]
Seems to work fine in Firefox. At some point in the future, we should redirect to the SVG.

**Pull request merged**